pkg/security/podsecuritypolicy/sysctl/mustmatchpatterns.go

```diff
@@ -38,48 +57,89 @@ var (
-func (s *mustMatchPatterns) Validate(pod *api.Pod) field.ErrorList {
-	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, s.validateAnnotation(pod, api.SysctlsPodAnnotationKey)...)
-	allErrs = append(allErrs, s.validateAnnotation(pod, api.UnsafeSysctlsPodAnnotationKey)...)
-	return allErrs
+func (s *mustMatchPatterns) divideSysctls(sysctls []api.Sysctl) (safeSysctls, unsafeSysctls []sysctlItem) {
+	for i, sysctl := range sysctls {
+		found := false
+		for _, ws := range s.safeWhitelist {
+			if sysctl.Name == ws {
+				found = true
+				break
+			}
+		}
+		if found {
+			safeSysctls = append(safeSysctls, sysctlItem{name: sysctl.Name, index: i})
+		} else {
+			unsafeSysctls = append(unsafeSysctls, sysctlItem{name: sysctl.Name, index: i})
+		}
+	}
+	return
 }
 
-func (s *mustMatchPatterns) validateAnnotation(pod *api.Pod, key string) field.ErrorList {
+// Validate ensures that the specified values fall within the range of the strategy.
+func (s *mustMatchPatterns) Validate(pod *api.Pod) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	fieldPath := field.NewPath("pod", "metadata", "annotations").Key(key)
-
-	sysctls, err := helper.SysctlsFromPodAnnotation(pod.Annotations[key])
-	if err != nil {
-		allErrs = append(allErrs, field.Invalid(fieldPath, pod.Annotations[key], err.Error()))
+	var sysctls []api.Sysctl
+	if pod.Spec.SecurityContext != nil {
+		sysctls = pod.Spec.SecurityContext.Sysctls
 	}
 
 	if len(sysctls) > 0 {
-		if len(s.patterns) == 0 {
-			allErrs = append(allErrs, field.Invalid(fieldPath, pod.Annotations[key], "sysctls are not allowed"))
-		} else {
-			for i, sysctl := range sysctls {
-				allErrs = append(allErrs, s.ValidateSysctl(sysctl.Name, fieldPath.Index(i))...)
+		fieldPath := field.NewPath("pod", "spec", "securityContext")
+
+		safeSysctls, unsafeSysctls := s.divideSysctls(sysctls)
+		if len(safeSysctls) > 0 {
+			// Is the sysctl forbidden?
+			if len(s.forbiddenSysctls) > 0 {
+				for _, sysctl := range safeSysctls {
+					// TODO(jchaloup): find the correct index i of the sysctl
+					allErrs = append(allErrs, s.ValidateForbiddenSysctl(sysctl.name, fieldPath.Child("sysctls").Index(sysctl.index))...)
+				}
+			}
+		}
+
+		if len(unsafeSysctls) > 0 {
+			// Is the sysctl allowed?
+			if len(s.allowedUnsafeSysctls) > 0 {
+				for _, sysctl := range unsafeSysctls {
+					// TODO(jchaloup): find the correct index i of the sysctl
+					allErrs = append(allErrs, s.ValidateSysctl(sysctl.name, fieldPath.Child("sysctls").Index(sysctl.index))...)
+				}
+			} else {
+				for _, sysctl := range unsafeSysctls {
+					allErrs = append(allErrs, field.ErrorList{field.Forbidden(fieldPath.Child("sysctls").Index(sysctl.index), fmt.Sprintf("unsafe sysctl %q is not allowed", sysctl.name))}...)
+				}
 			}
 		}
 	}
 
 	return allErrs
 }
 
+func (s *mustMatchPatterns) ValidateForbiddenSysctl(sysctlName string, fldPath *field.Path) field.ErrorList {
+	for _, s := range s.forbiddenSysctls {
+		if s[len(s)-1] == '*' {
```

Comment:
`strings.HasSuffix(s, "*")` and `strings.TrimSuffix(s, "*")` is more readable

Reply:
Replaced